### ISSUE TITLE
T12944 Core Lint Failure

### ISSUE THREAD

**Issue by mpickering** [tweag/ghc#78]
Ticket #75 was about how T12944 and one possible fix. However, the problem is in fact deeper.

Firstly, the reason why the problem was difficult to debug is that it manifested itself whilst the simplifier was rebuilding an expression. As such, the intermediate expression was never printed out in core between simplification steps, it does everything in one go almost. 

In particular, the optimisations in `rebuildCase` were hiding the problem.

So, I turned off all the optimisations in `rebuildCase` so the case actually got rebuilt and there was still a core lint error. This time from the new check I just added which checks whether the usage information on a variable aligns with a variables actual usage. 

```
11218 *** Core Lint errors : in result of Simplifier ***                              
11219 <no location info>: warning:                                                    
11220     In the pattern of a case alternative: (Poly1 ipv_s1oJ[1] :: a_a1jp,         
11221                                                  ipv_s1oK[1] :: a_a1jp)         
11222     Weight of variable does agree with its usage                                
11223     ipv_s1oJ[1]                                                                 
11224     1                                                                           
11225     Usage: 0 
```

This errors says that `ipv_s1oJ` has weight `1` but was used `0` times. Not good! 

We realise that this eventually leads us to `refineDefaultAlts` which replaces a `DEFAULT` with a `DataAlt` and introduces some **new** variables. These are attempted to be scaled but in this case, they are scaled by `1` but then not used in the RHS. The correct scaling is surely `0` or `Omega` as we are replacing a wildcard match so whatever variables we introduce will not be used (immediately) in the RHS. 

This isn't the end of the story. Disabling `refineDefaultAlts` still leaves us with a core lint error.

**Comment by mpickering** [comment]
Consider the following example of the case of case transformation:

```
case_w (case_y[1] x of _ -> y[1]) of ....
==>
case_y[1] x of _ -> (case_w y[1] of ...) 
```

The transformation does not preserve linearity. In the first case y has usage [1] as declared in the case as it is used once in the only branch. Once the case of case transformation is performed, the RHS of the case becomes the scrutinee and is scaled by `w` and so the declared usage an actual usage do not match.

**Comment by mpickering** [comment]
To be more precise. It is `CaseElim` which is firing. 

`CaseElim` operates when

```
    -- 2b.  Turn the case into a let, if                                              
    --      a) it binds only the case-binder                                          
    --      b) unlifted case: the scrutinee is ok-for-speculation                     
    --           lifted case: the scrutinee is in HNF (or will later be demanded) 
    -- See Note [Case to let transformation]
```

With the referenced note

```
  Note [Case to let transformation]                                               
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                               
  If a case over a lifted type has a single alternative, and is being             
  used as a strict 'let' (all isDeadBinder bndrs), we may want to do              
  this transformation:                                                            
                                                                                  
      case e of r       ===>   let r = e in ...r...                               
        _ -> ...r...                                                              
                                                                                  
  We treat the unlifted and lifted cases separately:                              
                                                                                  
  * Unlifted case: 'e' satisfies exprOkForSpeculation                             
    (ok-for-spec is needed to satisfy the let/app invariant).                     
    This turns     case a +# b of r -> ...r...                                    
    into           let r = a +# b in ...r...                                      
    and thence     .....(a +# b)....                                              
                                                                                  
    However, if we have                                                           
        case indexArray# a i of r -> ...r...                                      
    we might like to do the same, and inline the (indexArray# a i).               
    But indexArray# is not okForSpeculation, so we don't build a let              
    in rebuildCase (lest it get floated *out*), so the inlining doesn't           
    happen either.  Annoying.                                                     
                                                                                  
  * Lifted case: we need to be sure that the expression is already                
    evaluated (exprIsHNF).  If it's not already evaluated                         
        - we risk losing exceptions, divergence or                                
          user-specified thunk-forcing                                            
        - even if 'e' is guaranteed to converge, we don't want to                 
          create a thunk (call by need) instead of evaluating it                  
          right away (call by value)                                              
                                                                                  
    However, we can turn the case into a /strict/ let if the 'r' is               
    used strictly in the body.  Then we won't lose divergence; and                
    we won't build a thunk because the let is strict.                             
    See also Note [Eliminating redundant seqs]                                    
                                                                                  
    NB: absentError satisfies exprIsHNF: see Note [aBSENT_ERROR_ID] in MkCore.    
    We want to turn                                                               
       case (absentError "foo") of r -> ...MkT r...                               
    into                                                                          
       let r = absentError "foo" in ...MkT r...
```

**Comment by mpickering** [comment]
What happens is that 

1.  `$WIntofLog` is inlined

```
5434 Inlining done: T12944.$WIntOfLog[ω]                                             
5435     Inlined fn:  \ (@ (poly :: * -> *))                                         
5436                    (@ a)                                                        
5437                    (dt_a1bA[1] [Occ=Once] :: poly a) ->                         
5438                    case dt_a1bA[1] of dt_X1bF[1] { __DEFAULT ->                 
5439                    T12944.IntOfLog[ω] @ poly @ a dt_X1bF[1]                     
5440                    }                                                            
5441     Cont:   ApplyToTy T12944.Poly1                                              
5442             ApplyToTy a                                                         
5443             ApplyToVal nodup (T12944.Poly1[ω]                                   
5444                                 @ a (GHC.Num.fromInteger[ω] @ a $dNum_a1cY[ω] lvl_s1fO[ω]))
5445             Stop[BoringCtxt] T12944.IntOfLog T12944.Poly1 a  
```

Then `knownBranch` creates a let binding 

```
let dt_X1bF[1] = Poly1[ω] @ a_a1cX dt_s1g3[ω] in
IntOfLog[ω] @ Poly1 @ a_a1cX dt_X1cF[1]
```

But then the results is scrutinised and the argument `dt_X1cY` is discarded. 

```
5287 simplExprC                                                                      
5288   case v_avM[ω] of ω { IntOfLog _ [Occ=Dead] ->                                 
5289   case case v'_avN[ω] of ω { IntOfLog p_avL[ω] [Occ=Once] ->                    
5290        $WIntOfLog[ω]                                                            
5291          @ Poly1                                                                
5292          @ a_a1cX                                                               
5293          ($cnegateV_a1di[ω] @ a_a1cX $dNum_a1cZ[ω] p_avL[ω])                    
5294        }                                                                        
5295   of ω                                                                          
5296   { IntOfLog _ [Occ=Dead] ->                                                    
5297   $WIntOfLog[ω]                                                                 
5298     @ Poly1                                                                     
5299     @ a_a1cX                                                                    
5300     (Poly1[ω]                                                                   
5301        @ a_a1cX (fromInteger[ω] @ a_a1cX $dNum_a1cZ[ω] lvl_s1fP[ω]))            
5302   }                                                                             
5303   }                                                                             
5304   Stop[BoringCtxt] IntOfLog Poly1 a_a1cX
```

Then we end up with a let binding with multiplicity `[1]` but without any usages. And hence, core lint complains.

**Comment by aspiwack** [comment]
Should this be closed by 7c829bbff4501bb4c9b07596deb0f19e3f30013e ?

Either way: please add this transformation in the Linear Core document at tweag/linear-types.

**Comment by mpickering** [comment]
No, that fix turns out not to be right. The commit message is correct though. So I will paste it here for the future.

-----------------------------

The simplifier creates let bindings under certain circumstances which
are then inserted later, at as high a level as possible.

However, we have to be somewhat careful here when it comes to linearity
as if we create a floating binding `x`

```
case_w (let x[1] = "Foo" in Qux x) of
  Qux _ -> "Bar"
```

Then the let will continue outwards

```
let x[1] = "Foo"
in "Bar"
```

So we get a linearity failure as the one usage of `x` is eliminated.
However, because the ambient context is an `Omega` context, we know that
we will use the scrutinee `Omega` times and hence all bindings inside it
`Omega` times as well. The failure was that we created a `[1]` binding
whilst inside this context.

We also have to be careful as if we have a `[1]` case

```
case_1 (let x[1] = "Foo" in Qux x) of
  Qux x -> x
```

then the binding maintains the correct linearity once it is floated rom
the case and KnownBranch is performed.

```
let x[1] = "Foo"
in x
```

Why do we do this scaling afterwards rather than when the binding is
created? It is possible the binding comes from a point deep inside the
expression. It wasn't clear to me that we know enough about the context
at the point we make the binding due to the `SimplCont` type. It might
be thread this information through to get it right at definition site.
For now, I leave warnings and this message to my future self.

-------

#75 and #74 are the same problem

**Comment by aspiwack** [comment]
So really, what is difficult is that the `case` can disappear before we actually perform the permutation between the `let` and the `case`. Which destroys the information of how much we ought to scale the multiplicity.

Is it possible to scale the emitted lets as we are performing the KnownBranch optimisation?

**Comment by mpickering** [comment]
That's not the problem. 

If we have 

```
case (case v of ...) of
   _ -> ...
```
The problem is that we first simplify the scrutinee which yields a simplified expression `e1` and some lets named `floats1`. We then continue rebuilding the continuation (the outer case) but we don't pass in `floats1` into this computation. Whilst rebuilding the continuation we may encounter a `Select` continuation which corresponds to a case, when we do we perform case-of-case. However, `floats1` is not around at this point to be scaled. 

So either `floats1` needs to be passed into the function which rebuilds a continuation or the rebuilding function needs to return a value which indicates how much `case-of-case` has been performed. In the first case, we could scale the floats at the precise point a case-of-case occurs. In the second case, we scale them once at the end.

**Comment by aspiwack** [comment]
Where are the floats kept? In the state of a monad?

**Comment by mpickering** [comment]
They are returned from the function which rebuilds the zipper and inserted by calls to `wrapFloats` at certain points.

**Comment by aspiwack** [comment]
(PS: I don't think you can scale them all at the end, because the lets in the floats may come from different places and may not pass the same number of case-of-case, hence should be scaled differently)

**Comment by aspiwack** [comment]
I can't quite wrap my head around this, can you point me to the functions which emit the floats and rebuild the zipper.

**Comment by mpickering** [comment]
Sure, `simplExprF` and `rebuild` are two of the key functions.

**Comment by mpickering** [comment]
If you want a specific example, look at what happens in the `knownCon` function.

**Comment by aspiwack** [comment]
So, `rebuild` takes a `SimpleEnv` as an argument and returns one (like a `State` monad). So does `knownCon`. What's wrong with scaling the `SimpleEnv` when rebuilding (or deleting) a case?

**Comment by mpickering** [comment]
It returns `SimplFloats` not a `SimplEnv` unless I am missing something.

**Comment by aspiwack** [comment]
There seem to be a number of problems discussed in this issue. Related to scaling of floats, inlining, and such. The alias-like binders (which should be merged soon) should solve this issue. I'll have a closer look when merged.

**Comment by aspiwack** [comment]
Fixed by 97c759f.
